tooltips: use tippy tooltips for elements advertising a keyboard shortcut.

Ddharmani3 opened the pull request:
<!-- Describe your pull request here.-->
use a modern tippy tooltip style for remaining tooltip elements advertising a keyboard shortcut.

Fixes: <!-- Issue link, or clear description.-->
#24311

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

- Gear Menu:
![image](https://user-images.githubusercontent.com/64723994/230759114-9667f065-cb56-40d4-8096-14ab726bbcbb.png)
    
    - with fallback placement:
    ![image](https://user-images.githubusercontent.com/64723994/230759126-e1f22cf3-8ce6-499a-9deb-742cc379d66e.png)
    
    - Dark there:
    ![image](https://user-images.githubusercontent.com/64723994/230759140-62b1560b-5f0a-4426-8a13-db8daa6b5e22.png)


- User List:
![image](https://user-images.githubusercontent.com/64723994/230759185-166f1602-b4d1-4505-822c-45c272546deb.png)

![image](https://user-images.githubusercontent.com/64723994/230759246-757f6978-221d-4541-8602-efa63fbef978.png)

- Search_open icon:
  ![image](https://user-images.githubusercontent.com/64723994/232673457-28679673-76fb-4d5e-9b10-85d6760df2aa.png)

- Search_close_button:
  ![image](https://user-images.githubusercontent.com/64723994/232673502-32ee8263-72ed-4f43-bd32-9eb44112493e.png)

- Search_query input field:
  ![image](https://user-images.githubusercontent.com/64723994/232673602-2a6e56d3-8f1f-4c32-b7a4-e3aab1c91211.png)
  
  - on fallback placement:
    ![image](https://user-images.githubusercontent.com/64723994/232673862-75fa90f2-fe3c-447f-85f6-a4536bcac85d.png)

- Search_close icon:
![image](https://user-images.githubusercontent.com/64723994/230759269-459564b2-47db-4dc5-b40f-90979c1b3bb6.png)

- Manage stream settings: 
  - create new stream button: 
     ![image](https://user-images.githubusercontent.com/64723994/232159720-3dbdccbb-8300-4135-8bff-9747597c5d55.png)

  - Toggle subscription:
    ![image](https://user-images.githubusercontent.com/64723994/234315104-5c57fef8-5759-4887-a621-98254e57ce7a.png)

  - View stream:
    ![image](https://user-images.githubusercontent.com/64723994/234315257-7efb302a-3ba4-4b4d-b968-cf91600c5f89.png)

- Message Expander/condenser:
![image](https://user-images.githubusercontent.com/64723994/230759436-9c989ffd-875c-4fa1-a8c1-220b9bc1c2c7.png)

- View user card:
![image](https://user-images.githubusercontent.com/64723994/230759452-f6976aa8-d283-4b6b-b194-fad573678825.png)

![image](https://user-images.githubusercontent.com/64723994/230759457-f182e37c-4651-41d3-b933-37780c08be3f.png)

- Correction of hotkey for `All direct messages` tooltip:
  ![image](https://user-images.githubusercontent.com/64723994/234315715-3f2dff90-1862-4acc-879f-ae5cf59741c7.png)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ddharmani3 commented:
@timabbott All the tooltips from #24399 are fixed as per feedback from you in This PR(#25047), #24967, and #24962.

All these PRs are ready for review now.

alya commented:
Let's have the "Create new stream" tooltip open downwards like all the other tooltips around (and like we have in `main`).

alya commented:
Looks good to me in manual testing other than that.

alya commented:
Could you please list all the delays on these tooltips? The ones in the upper right (gear menu, etc.) definitely feel like they are showing up too soon to me.

Ddharmani3 commented:
> Could you please list all the delays on these tooltips? The ones in the upper right (gear menu, user list) definitely feel like they are showing up too soon to me.

@alya, I saw your comments on https://github.com/zulip/zulip/pull/25047(This PR), https://github.com/zulip/zulip/pull/24967, and https://github.com/zulip/zulip/pull/24962. In all of these PRs, every tooltip has the same delay, which is set to LONG_HOVER_DELAY as defined in tippyjs.js:
```
const LONG_HOVER_DELAY = [750, 20];
```

Ddharmani3 commented:
@alya, Even with a 750ms delay, the tooltips may still appear too quickly and feel sudden as they just pops up. To make the transition smoother, we can add a fade animation like below:
without fade animation:

https://user-images.githubusercontent.com/64723994/231171370-5af55164-5f3c-44f1-9a0e-d0abfde39828.mp4

with fade animation:

https://user-images.githubusercontent.com/64723994/231171459-5f9450b4-07bd-4c95-b39f-5fc7d7f997bb.mp4

alya commented:
I think for consistency, we should avoid adding animations on a handful of tooltips and not others.

Ddharmani3 commented:
> Let's have the "Create new stream" tooltip open downwards like all the other tooltips around (and like we have in `main`).

@amanagr updated the placement of  `create_stream_plus_button`. This PR is ready for review now.

amanagr commented on web/templates/navbar.hbs:
```diff
@@ -35,7 +35,7 @@
                                 <span class="search_icon search_open" ><i class="zulip-icon zulip-icon-search"></i></span>
                                 <input class="search-query input-block-level home-page-input" id="search_query" type="text" placeholder="{{t 'Search' }}"
-                                  autocomplete="off" aria-label="{{t 'Search' }}" title="{{t 'Search' }} (/)"/>
+                                  autocomplete="off"/>
```
Did someone mention that we should remove this ?

Ddharmani3 replied:
Since there was no discussion about this particular tooltip, I included it in the PR for feedback. Should I remove the bootstrap tooltip, given that the tippy tooltip with the same content is already displayed on the `search_closed` icon?

amanagr replied:
<img width="930" alt="Screenshot 2023-04-16 at 5 30 40 PM" src="https://user-images.githubusercontent.com/25124304/232308400-6cb9457c-5430-4abd-8404-f6152aaa2868.png">

@alya this code triggers a default browser tooltip which appears on a long hover on the search bar. I think It doesn't hurt to keep this, since the search bar is quite wide. What do you think?

alya replied:
> Since there was no discussion about this particular tooltip, I included it in the PR for feedback. 

@Ddharmani3 If you are including something for feedback, it is essential to call out the question in the PR description. Certainly, I will never know that there is such a question if it's only revealed by a careful review of the code.

alya replied:
This is noted specifically in the PR template:

>  Calls out remaining decisions and concerns.

alya replied:
Let's have a patter of always using the modern Tippy styling for tooltips that include keyboard shortcuts. So we can replace this long-hover tooltip with a Tippy one.

amanagr commented:
LGTM, other than us removing `title` from search bar without any discussion.

amanagr commented:
While we are here, we should also add tooltips to `close` and `search` icons in this screenshot.
<img width="930" alt="Screenshot 2023-04-16 at 5 29 26 PM" src="https://user-images.githubusercontent.com/25124304/232308367-3c63d127-095f-442a-ba80-969bc87c94d0.png">

alya commented:
> While we are here, we should also add tooltips to `close` and `search` icons in this screenshot. <img alt="Screenshot 2023-04-16 at 5 29 26 PM" width="930" src="https://user-images.githubusercontent.com/25124304/232308367-3c63d127-095f-442a-ba80-969bc87c94d0.png">

Yes, good call!

alya commented:
@Ddharmani3 Please include the additional screenshots in the PR description once you've had a chance to update the PR.

Ddharmani3 commented:
Updated the PR to address feedback. Also added updated screenshots in the PR description. This PR is ready for review now.

alya commented:
Cool, thanks!

@timabbott I'm marking this for integration review, but let's take a look at the delays together when you are looking this over. I definitely want a longer delay on the search bar tooltip, and possibly others as well.

Ddharmani3 commented:
@timabbott, I have resolved the conflicts and Zulipbot has also removed the `has conflict` label. I am not sure why PR is still showing merge conflicts.

Ddharmani3 commented:
@amanagr, I have shifted the logic to prevent a second appearance of the view user card tooltip to tippy-zulip-delayed-tooltip, I know this is not the right approach to do this but I have read the [thread on CZO ](https://chat.zulip.org/#narrow/stream/6-frontend/topic/tippy.20sticking.20around/near/1551038) about @palashb01 working on this issue to solve it for all the tooltips and I think this will not be needed then.

Your feedback on this?

palashb01 commented:
Hey @Ddharmani3 , I have already created the PR to prevent the second appearance of the `view-user-card` tooltip, and it was merged yesterday :)  So, I think you no longer need to make changes for this part.

Ddharmani3 commented:
> Hey @Ddharmani3 , I have already created the PR to prevent the second appearance of the `view-user-card` tooltip, and it was merged yesterday :) So, I think you no longer need to make changes for this part. #25147

@palashb01 I know that you have created PR that got merged which resulted in conflicts for this PR as this PR removes `.view_user_card_tooltip` and uses a general `.tippy-zulip-delayed-tooltip` for view user card tooltips. Hence to resolve the merge conflicts I shifted the logic used by you in #25147 to `.tippy-zulip-delayed-tooltip` which can be removed in the PR where you will fix this issue for all tooltips as mentioned by you [here](https://github.com/zulip/zulip/pull/25179#issuecomment-1514524802).

alya commented:
I'm not sure why this has a review request for me right now?

timabbott commented:
The tooltips seem to disappear immediately after appearing when testing this -- can you check you didn't break something rebasing this?

Ddharmani3 commented:
@timabbott fixed the PR. I kept the existing tippy for `.view_user_card_tooltip` instead of using `.tippy-zulip-delayed-tooltip` as `.view_user_card_tooltip` tippy has the logic for preventing a second appearance of the tooltip that was implemented in PR #25147. `.view_user_card_tooltip` now uses the modern hotkey tippy tooltip style.

alya commented:
Let's make the "Unsubscribe" button use the extra long delay, if you haven't done so yet, as the tooltip just adds info about the keyboard shortcut on that one.

Ddharmani3 commented:
> Let's make the "Unsubscribe" button use the extra long delay, if you haven't done so yet, as the tooltip just adds info about the keyboard shortcut on that one.

I have used EXTRA_LONG_HOVER_DELAY for `.tippy-zulip-delayed-tooltip` so every tooltip mentioned above has EXTRA_LONG_HOVER_DELAY except `.view_user_card_tooltip`.

alya commented:
Following the rule of thumb @timabbott noted in the EXTRA_LONG_HOVER_DELAY comment, only the "Unsubscribe" button should have the EXTRA_LONG_HOVER_DELAY. The rest should all use LONG_HOVER_DELAY.

Ddharmani3 commented:
> Following the rule of thumb @timabbott noted in the EXTRA_LONG_HOVER_DELAY comment, only the "Unsubscribe" button should have the EXTRA_LONG_HOVER_DELAY. The rest should all use LONG_HOVER_DELAY.

@alya, Updated the PR. Added a new tippy tooltip for the "Unsubscribe" button with  `.toggle-subscription-tooltip` with EXTRA_LONG_HOVER_DELAY and placement bottom.

Ddharmani3 commented:
@alya corrected the subscription toggle tooltip and view stream tooltip to read `Shift + S` and `Shift + V` respectively.
Also corrected the hotkey for the `All direct messages` tooltip to be `Shift + P`.

Updated the screenshots in the PR description.

This PR is ready for review now.

alya commented:
Thanks!

After chatting with @timabbott , let's drop the search tooltip that shows up in the middle of the search bar. The tooltip on the magnifying glass should be enough.

Otherwise looks good to me in screenshots and manual testing.

Ddharmani3 commented:
> After chatting with @timabbott , let's drop the search tooltip that shows up in the middle of the search bar. The tooltip on the magnifying glass should be enough.

Updated the PR to remove this. This PR is ready for review now.

timabbott commented:
I rebased this, made a couple small changes to strings, added `trigger: "mouseenter"` for the new tooltip class, since I noticed the default of "mouseenter focus" being a problem for the gear menu, and I think none of the tooltips using it want the on-focus behavior. With those changes, I've marked this to merge once CI passes.

I opened the release goal https://github.com/zulip/zulip/issues/25324 for a follow-up that should be straightforward but I expect to be of significant importance.

Thanks for all the work on this @Ddharmani3!

timabbott commented:
Another potential follow-up is that there's only 17 HTML `title` tooltips remaining outside settings in the app (another ~40 exist in settings), and it might be worth changing the rest of them to Tippy. The main ones I'd include in 7.0 if done are `recipient_row` and `edited_notice`, as the others are pretty subtle; probably each should be its own PR.

```
$ git grep 'title="' web/templates/ | grep -v settings
web/templates/draft.hbs:                <div class="recipient_row_date" title="{{t 'Last modified'}}">{{ time_stamp }}</div>
web/templates/draft.hbs:                <div class="recipient_row_date" title="{{t 'Last modified'}}">{{ time_stamp }}</div>
web/templates/draft.hbs:                    <div class="message_content rendered_markdown restore-draft restore-overlay-message" title="{{t 'Restore draft' }}">{{rendered_markdown content}}</div>
web/templates/edited_notice.hbs:<div class="message_edit_notice auto-select" title="{{t 'Edited ({last_edit_timestr})' }}">
web/templates/edited_notice.hbs:<div class="message_edit_notice auto-select" title="{{t 'Moved ({last_edit_timestr})' }}">
web/templates/edited_notice.hbs:<div class="message_edit_notice auto-select" title="{{t 'Edited ({last_edit_timestr})' }}">
web/templates/emoji_showcase.hbs:    <div class="emoji-canonical-name" title="{{name}}">{{name}}</div>
web/templates/gear_menu.hbs:            <li title="{{t 'Grant Zulip the Kerberos tickets needed to run your Zephyr mirror via Webathena' }}" id="webathena_login_menu" role="presentation">
web/templates/keyboard_shortcuts.hbs:                          title=":thumbs_up:"/>
web/templates/recipient_row.hbs:          title="{{t 'Narrow to stream "{display_recipient}"' }}">
web/templates/recipient_row.hbs:              title="{{t 'Narrow to stream "{display_recipient}", topic "{topic}"' }}">
web/templates/recipient_row.hbs:          title="{{t 'Narrow to your direct messages with {display_reply_to}' }}">
web/templates/scheduled_message.hbs:                        <div class="message_content rendered_markdown restore-overlay-message" title="{{t 'Edit or reschedule message' }}">{{rendered_markdown rendered_content}}</div>
web/templates/stream_sidebar_row.hbs:            <a href="{{url}}" title="{{name}}" class="stream-name">{{name}}</a>
web/templates/topic_list_item.hbs:        <a href='{{url}}' class="topic-name" title="{{topic_name}}">
web/templates/typeahead_list_item.hbs:  title="{{t 'You are not currently subscribed to this stream.' }}"></span>
web/templates/user_group_info_popover_content.hbs:                <span class="user_circle {{user_circle_class}}  popover_user_presence" title="{{user_last_seen_time_status}}"></span>

```

Ddharmani3 commented:
> Another potential follow-up is that there's only 17 HTML `title` tooltips remaining outside settings in the app (another ~40 exist in settings), and it might be worth changing the rest of them to Tippy. The main ones I'd include in 7.0 if done are `recipient_row` and `edited_notice`, as the others are pretty subtle; probably each should be its own PR.

I would love to work on follow-up PR, too. I will create follow-up PR soon.
